### PR TITLE
Detect actioned notes and pass additional parameters to onboarding URL

### PR DESCRIPTION
**Pull request by jrodger**
Closes https://github.com/Automattic/woocommerce-payments/issues/998

#### Changes proposed in this Pull Request

* If any WCPay admin notes have been actioned, add a list of them to the state data we pass through the onboarding flow

#### Testing instructions

* Requires some server-side changes, so this is best tested along with 410-gh-woocommerce-payments-server using the test instructions from that PR. 

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

### CODE REVIEW

File: includes/class-wc-payments-account.php
Comment (dechov) [inline includes/class-wc-payments-account.php:660]: Can use `get_notes_with_name` from the data store, as seen in [this post](https://developer.woocommerce.com/2020/10/16/using-the-admin-notes-inbox-in-woocommerce/).

File: includes/class-wc-payments-account.php
Comment (dechov) [inline includes/class-wc-payments-account.php:661]: I _think_ we could/should keep using this name – will be deprecated soon but isn't yet (pbIJXs-qi-p2).

File: includes/class-wc-payments-account.php
Comment (jrodger) [reply includes/class-wc-payments-account.php]: Ah ha, great. Thanks!

File: includes/class-wc-payments-account.php
Comment (jrodger) [reply includes/class-wc-payments-account.php]: Thanks for the link. I've confirmed that these classes are available on our earliest supported WC version (4.0). I'll create an issue so that we can write some version specific code for WC 4.8 later.

**Comment by jrodger** [comment]
The latest changes are more generic and will send a list of wcpay admin note names along in the onboarding state. Looking them up is done in memory, but we do have another option, using the `woocommerce_note_where_clauses` filter:
```
$add_like_clause = function( $where_clause ) {
	return $where_clause . " AND name like 'wcpay-promo-%'";
};

add_filter( 'woocommerce_note_where_clauses', $add_like_clause );

$notes = $data_store->get_notes( [
	'status'     => [ WC_Admin_Note::E_WC_ADMIN_NOTE_ACTIONED ],
	'is_deleted' => false
] );

remove_filter( 'woocommerce_note_where_clauses', $add_like_clause );
```

I'd appreciate any opinions on which approach feels the simplest / most robust.

File: includes/class-wc-payments-account.php
Comment (kalessil) [inline includes/class-wc-payments-account.php:673]: Can we move this into try-clause, so we get something like `$notes = WC_Data_Store::load( 'admin-note' )->get_notes(...)`. It'll greatly simplify the method as we can omit the if below and just force-cast notes to array in foreach?

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (kalessil) [inline includes/wc-payment-api/class-wc-payments-api-client.php:650]: I also think we can declare the $additional type as `array` without breaking any compatibility.

File: includes/class-wc-payments-account.php
Comment (kalessil) [inline includes/class-wc-payments-account.php:655]: Can we also declare a return type here (`private function get_actioned_notes(): array`)?

**Comment by kalessil** [comment]
@jrodger: I'd go for using `woocommerce_note_where_clauses` solution:
- then limit can be reduced
- fewer data will be trasferred between PHP and DB

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (jrodger) [reply includes/wc-payment-api/class-wc-payments-api-client.php:650]: Good plan, I always forget we have the necessary minimum PHP version for this. Please keep prompting when I miss these!

File: includes/class-wc-payments-account.php
Comment (jrodger) [reply includes/class-wc-payments-account.php:655]: We can, thanks.

File: includes/class-wc-payments-account.php
Comment (jrodger) [reply includes/class-wc-payments-account.php:673]: > @jrodger: I'd go for using `woocommerce_note_where_clauses` solution:
> 
> * then limit can be reduced (we limit exactly promo notes)
> * fewer data will be transferred between PHP and DB (resources-friendly)

It's certainly a better solution for those reasons. I was a bit worried about messing around with the SQL query as a string, but I think that's outweighed by your points there.

**Comment by jrodger** [comment]
> @jrodger: I'd go for using `woocommerce_note_where_clauses` solution:
> 
> * then limit can be reduced (we limit exactly promo notes)
> * fewer data will be transferred between PHP and DB (resources-friendly)

It's certainly a better solution for those reasons. I was a bit worried about messing around with the SQL query as a string, but I think that's outweighed by your points there.

File: includes/class-wc-payments-account.php
Comment (jrodger) [reply includes/class-wc-payments-account.php:673]: If we go with the where clause approach then we won't need to foreach loop. However, we'd still want to cast the array before it was returned.

Were you thinking something like `return (array)$wcpay_promo_notes;`?

Comment (kalessil) [reply includes/class-wc-payments-account.php:673]: Indeed. I was thinking about `foreach ( (array) $notes as ...)`

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (kalessil) [reply includes/wc-payment-api/class-wc-payments-api-client.php:650]: Gladly =)

**Comment by jrodger** [comment]
I'd like to get some more unit tests around this, but the refactoring to get that working is perhaps a little out of scope for this PR.

**Comment by kalessil** [comment]
The changes are looking awesome, thank you @jrodger. Regarding the test: do you mean that you'll be adding more tests here?

**Comment by jrodger** [comment]
> The changes are looking awesome, thank you @jrodger. Regarding the tests: do you mean that you'll be adding more tests here?

There aren't currently any unit tests for the method we're updating, adding them would require some refactoring due to the `exit` statements (and other things) in the code. So I'm not planning to add tests as part of this PR (as much as I would like to).

**Review by kalessil — APPROVED** [review]
Looks like the code sniffer found some violations, but the rest LGTM. :shipit:

**Comment by jrodger** [comment]
The PHPCS warnings have been fixed, this should be ready to go but let's get the server portion reviewed and deployed first. I'll leave this as `Needs Review` for now since testing the server part requires testing this as well.

**Comment by jrodger** [comment]
The server changes have been deployed, so this is ready for a final review.

**Review by luizreis — APPROVED** [review]
Looks good! I tested this PR in WC `4.0` and `4.7`, with and without the note and all cases worked perfectly. 🎉 

I'm approving this, but we need to fix the `phpcs` sniff violation before merging it.

File: includes/class-wc-payments-account.php
Comment (jrodger) [reply includes/class-wc-payments-account.php]: Missing linting violations is becoming a theme for me on this PR. Thanks, fixed in https://github.com/Automattic/woocommerce-payments/pull/1044/commits/35072c8f03607f0edcdde4b305cfbdba26624730

**Pull request merged**
